src/bin/pg_autoctl/ipaddr.c

```diff
@@ -244,3 +302,226 @@ isTestEnv(void)
 	const char *socketDir = getenv("PG_REGRESS_SOCK_DIR");
 	return socketDir != NULL && socketDir[0] == '\0';
 }
+
+
+/*
+ * From /Users/dim/dev/PostgreSQL/postgresql/src/backend/libpq/hba.c
+ */
+static bool
+ipv4eq(struct sockaddr_in *a, struct sockaddr_in *b)
+{
+	return (a->sin_addr.s_addr == b->sin_addr.s_addr);
+}
+
+
+/*
+ * From /Users/dim/dev/PostgreSQL/postgresql/src/backend/libpq/hba.c
+ */
+static bool
+ipv6eq(struct sockaddr_in6 *a, struct sockaddr_in6 *b)
+{
+	int			i;
+
+	for (i = 0; i < 16; i++)
+		if (a->sin6_addr.s6_addr[i] != b->sin6_addr.s6_addr[i])
+			return false;
+
+	return true;
+}
+
+
+/*
+ * findHostnameLocalAddress does a reverse DNS lookup given an hostname
+ * (--nodename), and if the DNS lookup fails or doesn't return any local IP
+ * address, then returns false.
+ */
+bool
+findHostnameLocalAddress(const char *hostname, char *localIpAddress, int size)
+{
+	int error;
+    struct addrinfo *dns_lookup_addr;
+	struct addrinfo *dns_addr;
+	struct ifaddrs *ifaddr, *ifa;
+
+	error = getaddrinfo(hostname, NULL, 0, &dns_lookup_addr);
+	if (error != 0)
+	{
+		log_warn("Failed to resolve DNS name \"%s\": %s",
+				 hostname, gai_strerror(error));
+		return false;
+	}
+
+	/*
+	 * Loop over DNS results for the given hostname. Filter out loopback
+	 * devices, and for each IP address given by the look-up, check if we
+	 * have a corresponding local interface bound to the IP address.
+	 */
+
+	if (getifaddrs(&ifaddr) == -1)
+	{
+		log_warn("Failed to get the list of local network inferfaces: %s",
+				  strerror(errno));
+		return false;
+	}
+
+	/*
+	 * Compare both addresses list (dns lookup and list of interface
+	 * addresses) in a nested loop fashion: lists are not sorted, and we
+	 * expect something like a dozen entry per list anyway.
+	 */
+	for (dns_addr=dns_lookup_addr; dns_addr; dns_addr = dns_addr->ai_next)
+	{
+		/*
+		 * Skip loopback addresses, we want to be able to connect to
+		 * hostname from other nodes.
+		 */
+		if (!(dns_addr->ai_flags & IFF_LOOPBACK))
+		{
+			for (ifa = ifaddr; ifa; ifa = ifa->ifa_next)
+			{
+				/*
+				 * TODO: implement ipv6 support
+				 */
+				if (ifa->ifa_addr->sa_family == AF_INET
+					&& dns_addr->ai_family == AF_INET)
+				{
+					struct sockaddr_in *ip =
+						(struct sockaddr_in *) ifa->ifa_addr;
+
+					if (ipv4eq(ip, (struct sockaddr_in *) dns_addr->ai_addr))
```

Comment:
wonder here if it would be more readable by either moving this check into the if-statement above, or replacing with a `continue;` on failure. Again to remove a bit of the indenting.

Same for the block below.

Reply:
Done.